### PR TITLE
Add Upsert Search Attributes

### PR DESCRIPTION
**Pull request by vancexu**
This PR add new feature UpsertSearchAttributes to allow use update search attributes inside workflow implementation. 

TODO in separate PR is add search attributes to workflow info, so that user can easily get such info.

### CODE REVIEW

File: src/main/java/com/uber/cadence/workflow/Workflow.java [hunk 1]
```diff
@@ -1182,6 +1183,39 @@ public static Logger getLogger(String name) {
...
+   *
+   * <pre><code>
+   *     Map<String, Object> attr1 = new HashMap<>();
+   *     attr1.put("CustomIntField", 1);
+   *     attr1.put("CustomBoolField", true);
+   *     Workflow.upsertSearchAttributes(attr1);
+   *
+   *     Map<String, Object> attr2 = new HashMap<>();
+   *     attr2.put("CustomIntField", 2);
+   *     attr2.put("CustomKeywordField", "key");
```
Comment (meiliang86) [inline src/main/java/com/uber/cadence/workflow/Workflow.java]: value should be seattle?

File: src/test/java/com/uber/cadence/workflow/WorkflowTest.java [hunk 2]
```diff
@@ -183,7 +183,7 @@ protected void failed(Throwable e, Description description) {
         }
       };
 
-  @Parameter public boolean useExternalService;
+  @Parameter public boolean useExternalService = true;
```
Comment (meiliang86) [inline src/test/java/com/uber/cadence/workflow/WorkflowTest.java]: revert this before checking in.

File: src/main/java/com/uber/cadence/internal/replay/ClockDecisionContext.java [hunk 3]
```diff
@@ -217,6 +218,14 @@ private void timerCancelled(long startEventId, Exception reason) {
     return mutableSideEffectHandler.handle(id, converter, func);
   }
 
+  void upsertSearchAttributes(SearchAttributes searchAttributes) {
+    decisions.upsertSearchAttributes(searchAttributes);
+  }
+
+  void handleUpsertSearchAttributes(HistoryEvent event) {
```
[the remark this answers is not in the document]
Comment (vancexu) [reply src/main/java/com/uber/cadence/internal/replay/ClockDecisionContext.java]: done

File: src/main/java/com/uber/cadence/workflow/Workflow.java (hunk 1 above)
Comment (vancexu) [reply src/main/java/com/uber/cadence/workflow/Workflow.java]: yep, thanks

File: src/test/java/com/uber/cadence/workflow/WorkflowTest.java (hunk 2 above)
Comment (vancexu) [reply src/test/java/com/uber/cadence/workflow/WorkflowTest.java]: thanks for the catch. done

**Comment by vancexu** [comment]
> Have you updated the unit test framework?

Let me add test framework support in another PR.

**Pull request merged**